src/particles/elements/Aperture.H

```diff
@@ -0,0 +1,119 @@
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Chad Mitchell, Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef IMPACTX_APERTURE_H
+#define IMPACTX_APERTURE_H
+
+
+#include "particles/ImpactXParticleContainer.H"
+#include "mixin/beamoptic.H"
+#include "mixin/thin.H"
+#include "mixin/nofinalize.H"
+
+#include <ablastr/particles/IndexHandling.H>
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+
+#include <cmath>
+
+namespace impactx
+{
+    struct Aperture
+    : public elements::BeamOptic<Aperture>,
+      public elements::Thin,
+      public elements::NoFinalize
+    {
+        static constexpr auto name = "Aperture";
+        using PType = ImpactXParticleContainer::ParticleType;
+
+        /** A thin collimator element that applies a transverse aperture boundary.
+         *  Particles outside the boundary are considered lost.
+         *
+         * @param shape aperture shape (0 rectangular, 1 elliptical)
+         * @param xmax maximum value of horizontal coordinate (m)
+         * @param ymax maximum value of vertical coordinate (m)
+         */
+        Aperture( int const shape,
+                   amrex::ParticleReal const xmax,
+                   amrex::ParticleReal const ymax )
+        : m_shape(shape), m_xmax(xmax), m_ymax(ymax)
+        {
+        }
+
+        /** Push all particles */
+        using BeamOptic::operator();
+
+        /** This is an aperture functor, so that a variable of this type can be used like an
+         *  aperture function.
+         *
+         * @param p Particle AoS data for positions and cpu/id
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         * @param refpart reference particle (unused)
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (
+                PType& AMREX_RESTRICT p,
+                [[maybe_unused]] amrex::ParticleReal & AMREX_RESTRICT px,
+                [[maybe_unused]] amrex::ParticleReal & AMREX_RESTRICT py,
+                [[maybe_unused]] amrex::ParticleReal & AMREX_RESTRICT pt,
+                [[maybe_unused]] RefPart const & refpart) const {
+
+            using namespace amrex::literals; // for _rt and _prt
+
+            // access AoS data such as positions and cpu/id
+            amrex::ParticleReal const x = p.pos(RealAoS::x);
+            amrex::ParticleReal const y = p.pos(RealAoS::y);
+            int const id = p.id();
+            //uint64_t const global_id = ablastr::particles::localIDtoGlobal(p.id(), p.cpu());
+
+            // scale horizontal and vertical coordinates
+            amrex::ParticleReal const u = x/m_xmax;
+            amrex::ParticleReal const v = y/m_ymax;
+
+            // compare against the aperture boundary
+
+            switch(m_shape) {
+
+           // rectangular aperture (default)
```

Comment:
Not super important yet, but since only one option can be true at the same time, we potentially want to move this switch decision out of kernel code, where it increases register pressure, and into a host-side decision.

For this, we could write a general aperture base class and derive it into special rectangular and elliptical classes.

Anyway, I think the part here is too small to warrant this just yet and this will be fine.